cjdns-bytes/src/utils.rs

```diff
@@ -1,150 +1,224 @@
 //! Parsing and serialization helpers.
+pub use reader::{Reader, SizePredicate};
+pub use writer::Writer;
 
-use std::convert::TryFrom;
+mod reader {
+    use std::convert::TryFrom;
 
-type Result<T> = std::result::Result<T, ()>;
+    type Result<T> = std::result::Result<T, InsufficientBuffer>;
 
-pub struct Reader<'a>(&'a [u8]);
-pub struct Writer(Vec<u8>);
+    pub struct Reader<'a>(&'a [u8]);
+    #[derive(Debug)]
+    pub struct InsufficientBuffer;
+    pub struct ReadError;
 
-impl<'a> Reader<'a> {
-    pub fn new(data: &'a [u8]) -> Self {
-        Reader(data)
+    pub enum SizePredicate {
```

Comment:
Predicate is a boolean expression. This is not a predicate.

Reply:
`SizeOrdering`?

Reply:
Maybe, `ExpectedSize`

Reply:
Or, `AssertSize`

Reply:
done